File: python_backend/game_logic/ranges/king/get_king_actual_moves.py
from game_logic.threatArea.top.sqr_under_attack import sqr_under_attack
from game_logic.threatArea.get_hypo_boards import get_hypo_boards
from game_logic.test_objects.get_standard_range_defs import get_standard_range_defs
from game_logic.test_objects.get_standard_id_dict import get_standard_id_dict
from game_logic.test_objects.sample_board_dicts import sample_board_dicts
from game_logic.printers.print_board import print_board


def get_king_actual_moves(board, king_loc, init_range, color, range_defs, id_dict):
    """filter the king's initial moves to get the actual ones"""
    boards = get_hypo_boards(board, king_loc, init_range, color)
    actual_moves = []
    for hypo_k_loc, hypo_board in boards.items():
        if not sqr_under_attack(hypo_board, hypo_k_loc, color, range_defs, id_dict):
            actual_moves.append(hypo_k_loc)
    return actual_moves


if __name__ == "__main__":
    range_defs = get_standard_range_defs()
    id_dict = get_standard_id_dict()

    board = sample_board_dicts['super_checkmate_impossible_example']
    actual_moves = get_king_actual_moves(board, (6, 6), [(5, 5), (6, 5), (7, 5), (5, 6), (7, 6), (5, 7), (6, 7), (7, 7)],
                                         "W", range_defs, id_dict)
    print_board(board, heading="super_checkmate_impossible_example", highlights=actual_moves)
    print('')

    board = sample_board_dicts['king_range_test']
    actual_moves = get_king_actual_moves(board, (4, 3), [(3, 2), (4, 2), (3, 3), (5, 3), (3, 4), (5, 4)],
                                         "W", range_defs, id_dict)
    print_board(board, heading="king_range_test", highlights=actual_moves)
    print('')

File: app (1)/src/game_logic/ranges/king/getKingActualMoves.js
import {sqrUnderAttack} from "../../threatArea/top/sqrUnderAttack";
import {getHypoBoards} from "../../threatArea/getHypoBoards";
import {mapListXyToRf} from "../../coordType/mapListXyToRf";
import { rfToXy } from "../../coordType/crdCnvrt";


export function getKingActualMoves(board, kingLoc, initRange, color, pieceDefs, idDict) {
    /*filter the king's initial moves to get the actual ones**/
    var boards = getHypoBoards(board, kingLoc, initRange, color)
    var actualMoves = []
    for (var [hypoKLoc, hypoBoard] of Object.entries(boards)) {
        if (! sqrUnderAttack(hypoBoard, rfToXy(hypoKLoc), color, pieceDefs, idDict)) {
            actualMoves.push(rfToXy(hypoKLoc))
        }
    }
    return actualMoves
}

// module.exports = getKingActualMoves;

/** for node.js
if (require.main === module) {
    
    import {JsonRecords} from "../../JsonRecords/JsonRecords";
    import {SpecialMoves} from "../specialMoves/SpecialMoves";
    import {getStandardPieceDefs} from "../../testObjects/getStandardPieceDefs";
    import {getStandardIdDict} from "../../testObjects/getStandardIdDict";
    import {sampleBoardDicts} from "../../testObjects/sampleBoardDicts";
    import {printBoard} from "../../printers/printBoard";
    import {initPawnIds} from "../../JsonRecords/initPawnIds";
    import {initEmptyRanges} from "../../apiHelpers/initEmptyRanges";
    import {dataDict} from "../../testObjects/dataDicts/dd1";
    
    var board, records, jsonRecords, specialMoves, actualMoves;
    var pieceDefs = initEmptyRanges(getStandardPieceDefs())
    var idDict = getStandardIdDict()

    //king range test
    board = sampleBoardDicts["king_range_test"];
    records = dataDict["king_range_test"]["json"]
    jsonRecords =  new JsonRecords(initPawnIds(records, board))
    specialMoves = new SpecialMoves()
    actualMoves = getKingActualMoves(board, [4, 3], [[3, 2], [4, 2], [3, 3], [5, 3], [3, 4], [5, 4]], "W", pieceDefs, idDict)
    printBoard(board, "\nking_range_test", mapListXyToRf(actualMoves))

    //super_checkmate_impossible_example
    board = sampleBoardDicts['super_checkmate_impossible_example']
    records = dataDict["super_checkmate_impossible_example"]["json"]
    jsonRecords =  new JsonRecords(initPawnIds(records, board))
    specialMoves = new SpecialMoves()
    actualMoves = getKingActualMoves(board, [6, 6], [[5, 5], [5, 6], [5, 7], [6, 5], [6, 7], [7, 5], [7, 6], [7, 7]], "W", pieceDefs, idDict)
    printBoard(board, "super_checkmate_impossible_example", mapListXyToRf(actualMoves))

}
*/